Interop/Runtime/src/main/kotlin/kotlinx/cinterop/Utils.kt

```diff
@@ -69,30 +69,61 @@ public abstract class AutofreeScope : DeferScope(), NativePlacement {
     abstract override fun alloc(size: Long, align: Int): NativePointed
 }
 
+private const val ChunkSize = 4096
+
 public open class ArenaBase(private val parent: NativeFreeablePlacement = nativeHeap) : AutofreeScope() {
 
-    private var lastChunk: NativePointed? = null
+    private var lastLargeObjectsChunk: NativePointed? = null
+    private var lastSmallObjectsChunk: NativePointed? = null
+    private var offset = 0
 
-    final override fun alloc(size: Long, align: Int): NativePointed {
+    private fun allocLargeObject(size: Long, align: Int): NativePointed {
         // Reserve space for a pointer:
         val gapForPointer = maxOf(pointerSize, align)
 
         val chunk = parent.alloc(size = gapForPointer + size, align = gapForPointer)
-        nativeMemUtils.putNativePtr(chunk, lastChunk.rawPtr)
-        lastChunk = chunk
+        nativeMemUtils.putNativePtr(chunk, lastLargeObjectsChunk.rawPtr)
+        lastLargeObjectsChunk = chunk
         return interpretOpaquePointed(chunk.rawPtr + gapForPointer.toLong())
     }
 
+    final override fun alloc(size: Long, align: Int): NativePointed {
+        if (size >= ChunkSize / 2)
+            return allocLargeObject(size, align)
+
+        val startPtr = lastSmallObjectsChunk?.rawPtr?.toLong()
```

Comment:
Redundant boxing is possible here.

Reply:
Also chain of safe calls produce chain of null checks, while only first one is required.